File: src/test/cpp/datastructures/trees/binarysearchtrees/binarysearchtree/binary_search_tree_test.cpp
#include <gtest/gtest.h>
#include <cmath>
#include <vector>
#include <unordered_set>

#include "main/cpp/datastructures/trees/binarysearchtrees/binarysearchtree/BinarySearchTree.h"

#include "main/cpp/my_utils.h"

TEST(BinarySearchTreesTest, NoNodesDeletionTest) {
  BinarySearchTree<int> tree;
  ASSERT_FALSE(tree.remove(1));
}

TEST(BinarySearchTreesTest, Duplicates) {
  BinarySearchTree<int> tree;
  tree.insert(1);
  tree.insert(2);
  tree.insert(1);
  tree.insert(2);

  ASSERT_EQ(tree.numNodes(), 2);
}

TEST(BinarySearchTreesTest, DuplicatesDeletion) {
  BinarySearchTree<int> tree;
  tree.insert(1);
  tree.insert(2);
  tree.insert(1);
  ASSERT_TRUE(tree.remove(1));
  ASSERT_FALSE(tree.remove(1));
  ASSERT_FALSE(tree.remove(1));
  ASSERT_TRUE(tree.remove(2));

  ASSERT_EQ(tree.numNodes(), 0);
}

File: java/src/test/java/datastructures/trees/binarysearchtrees/BinarySearchTreeTest.java
package datastructures.trees.binarysearchtrees;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import datastructures.trees.binarysearchtrees.binarysearchtree.BinarySearchTree;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class BinarySearchTreeTest {

  private BinarySearchTree<Integer> tree;

  @BeforeEach
  public void setup() {
    tree = new BinarySearchTree<>();
  }

  @Test
  public void insertTest() {
    assertFalse(tree.contains(5));
    tree.insert(5);
    assertTrue(tree.contains(5));

    for (int i = 0; i < 100; i++) {
      tree.insert((int) (Math.random() * 100));
      assertTrue(tree.isValid());
    }
  }

  @Test
  public void deleteTest() {
    tree.insert(5);
    assertTrue(tree.contains(5));
    tree.delete(5);
    assertFalse(tree.contains(5));

    int[] values = new int[100];

    for (int i = 0; i < values.length; i++) {
      values[i] = (int) (Math.random() * 100);
      tree.insert(values[i]);
    }

    for (int i : values) {
      tree.delete(i);
      assertTrue(tree.isValid());
    }
  }
}
